imagedephi/redact/tiff.py

```diff
@@ -0,0 +1,128 @@
+from __future__ import annotations
+
+from collections.abc import Generator
+from itertools import chain
+from pathlib import Path
+from typing import TYPE_CHECKING, Type
+
+import tifftools
+import tifftools.constants
+
+from imagedephi.rules import FileFormat, MetadataTiffRule, RuleSet
+
+from .redaction_plan import FILE_EXTENSION_MAP, RedactionPlan
+
+if TYPE_CHECKING:
+    from tifftools.tifftools import IFD, TiffInfo
+
+
+class TiffBasedMetadataRedactionPlan(RedactionPlan):
+    tiff_info: TiffInfo
+
+    def __init__(self, tiff_info: TiffInfo, base_rules: RuleSet, override_rule_set: RuleSet | None):
+        self.tiff_info = tiff_info
+
+    @classmethod
+    def get_all_subclasses(cls) -> Generator[Type[TiffBasedMetadataRedactionPlan], None, None]:
+        for subclass in cls.__subclasses__():
+            yield from subclass.get_all_subclasses()
+            yield subclass
+
+    @classmethod
+    def build(
+        cls, image_path: Path, base_rules: RuleSet, override_rules: RuleSet | None = None
+    ) -> TiffBasedMetadataRedactionPlan:
+        file_extension = FILE_EXTENSION_MAP[image_path.suffix]
+        for redaction_plan_class in cls.get_all_subclasses():
+            if file_extension == redaction_plan_class.file_format:
+                tiff_info = tifftools.read_tiff(str(image_path))
+                return redaction_plan_class(tiff_info, base_rules, override_rules)
+        else:
+            raise Exception(f"File format for {image_path} not supported.")
```

Comment:
@brianhelba This is slightly different. Previously the `build` class method would do hard checks against members of the `FileFormat` enum. Now it traverses its subclasses. Because of the current class structure, it also needs to check all of its grandchildren, hence the addition of the `get_all_subclasses` method.

Reply:
This also meant that I had to unify the signatures of the `__init__` functions on `TiffMetadataRedactionPlan` and `SvsMetadataRedactionPlan`.